File: src/client/test/app/LineChartTest.js
/**
 * A test class for the echopoint.google.chart.LineChart client-side component.
 * Displays simple and multi value charts to test Google Chart API interaction.
 *
 * @author Rakesh 2008-08-08
 * @version $Id: LineChartTest.js,v 1.1.1.1 2010-04-01 09:47:33 perxi Exp $
 */
echopoint.test.LineChartTest = Core.extend(
{
  $construct: function( testArea )
  {
    testArea.add( this._createSimple() );
    testArea.add( this._createComplex() );
  },

  _createSimple: function()
  {
    var xdata = [ 30,60,70,90,95,110 ];
    var xmax = 120;

    var markers = new Array();
    markers[0] = new echopoint.google.chart.model.ShapeMarker( "o", "ff3333", 5 );

    var data = new echopoint.google.chart.model.ChartData( xdata, xmax );
    data.color = "00ff00";
    data.markers = markers;

    var title = new echopoint.google.chart.model.Title();
    title.add( "Simple Chart" );

    var labels = new Array();
    labels[0] = [ "0", "20", "40", "60", "80", "100" ];
    labels[1] = [ "0", "25", "50", "75", "100" ];

    var ranges = new Array();
    ranges[0] = new echopoint.google.chart.model.Range( 20, 125 );
    ranges[1] = new echopoint.google.chart.model.Range( 25, 150 );

    var rangeMarkers = new Array();
    rangeMarkers[0] = new echopoint.google.chart.model.RangeMarker( "r", "ff0000", 0.1, 0.11 );
    rangeMarkers[1] = new echopoint.google.chart.model.RangeMarker( "R", "0000ff", 0.1, 0.11 );

    return new echopoint.google.chart.LineChart(
    {
      renderId: "echopointUnitTestSimpleLineChart",
      styleName: "SimpleChart",
      data: [ data ],
      title: title,
      axisLabels: labels,
      axisRanges: ranges,
      rangeMarkers: rangeMarkers
    });
  },

  _createComplex: function()
  {
    var markers1 = new Array();
    markers1[0] = new echopoint.google.chart.model.ShapeMarker( "a", "00ff33", 7 );
    markers1[1] = new echopoint.google.chart.model.ShapeMarker( "c", "00ff33", 7 );
    markers1[2] = new echopoint.google.chart.model.ShapeMarker( "d", "00ff33", 7 );
    markers1[3] = new echopoint.google.chart.model.ShapeMarker( "o", "00ff33", 7 );
    markers1[4] = new echopoint.google.chart.model.ShapeMarker( "s", "00ff33", 7 );
    markers1[5] = new echopoint.google.chart.model.ShapeMarker( "tValue+7", "00ff33", 7 );
    markers1[6] = new echopoint.google.chart.model.ShapeMarker( "x", "00ff33", 7 );

    var data = new Array();
    data[0] = new echopoint.google.chart.model.ChartData(
        [ 0, 30, 60, 70, 90, 95, 100 ], 110 );
    data[0].ydata = [ 20,30,40,50,60,70,80 ];
    data[0].ymax = 100;
    data[0].legend = "First";
    data[0].markers = markers1;

    var markers2 = new Array();
    markers2[0] = new echopoint.google.chart.model.ShapeMarker( "x", "ff00ff", 7 );
    markers2[1] = new echopoint.google.chart.model.ShapeMarker( "s", "ff00ff", 7 );
    markers2[2] = new echopoint.google.chart.model.ShapeMarker( "o", "ff00ff", 7 );
    markers2[3] = new echopoint.google.chart.model.ShapeMarker( "d", "ff00ff", 7 );
    markers2[4] = new echopoint.google.chart.model.ShapeMarker( "c", "ff00ff", 7 );
    markers2[5] = new echopoint.google.chart.model.ShapeMarker( "a", "ff00ff", 7 );

    data[1] = new echopoint.google.chart.model.ChartData(
        [ 10,30,40,45,52 ], 60 );
    data[1].ydata = [ 100,90,40,20,10 ]
    data[1].ymax = 110;
    data[1].legend = "Second";
    data[1].markers = markers2;

    var markers3 = new Array();
    markers3[0] = new echopoint.google.chart.model.ShapeMarker( "x", "ff3300", 7 );
    markers3[1] = new echopoint.google.chart.model.ShapeMarker( "a", "ff3300", 7 );
    markers3[2] = new echopoint.google.chart.model.ShapeMarker( "o", "ff3300", 7 );
    markers3[3] = new echopoint.google.chart.model.ShapeMarker( "c", "ff3300", 7 );
    markers3[4] = new echopoint.google.chart.model.ShapeMarker( "t", "ff3300", 7 );

    data[2] = new echopoint.google.chart.model.ChartData( [ -1 ], 0 );
    data[2].ydata = [ 5,33,50,55,7 ];
    data[2].ymax = 60;
    data[2].legend = "Third";
    data[2].markers = markers3;

    var title = new echopoint.google.chart.model.Title( "Complex Chart" );
    title.add( "Multiple Line Title" );

    var labels = new Array();
    labels[0] = [ 0, 20, 40, 60, 80, 100 ];
    labels[1] = [ 0, 25, 50, 75, 100 ];
    labels[2] = [ "Min", "One Third", "Full" ];
    labels[3] = [ 0, 50, 100 ];

    var positions = new Array();
    positions[0] = [];
    positions[1] = [];
    positions[2] = [ 0, 3, 10 ]; // No idea what values to show, this does not work
    positions[3] = [];

    var lineStyles = new Array();
    lineStyles[0] = new echopoint.google.chart.model.LineStyle( 3, 6, 3 );
    lineStyles[1] = new echopoint.google.chart.model.LineStyle( 2, 4, 2 );
    lineStyles[2] = new echopoint.google.chart.model.LineStyle( 4 );

    var ranges = new Array();
    ranges[0] = new echopoint.google.chart.model.RangeMarker( "r", "e5ecf9", 0.35, 0.25 );
    ranges[1] = new echopoint.google.chart.model.RangeMarker( "R", "a0bae9", 0.35, 0.25 );


    var fillArea = new Array();
    fillArea[0] = new echopoint.google.chart.model.FillArea( "b", "224499", 0, 1 );
    fillArea[1] = new echopoint.google.chart.model.FillArea( "b", "76a4fb", 1, 2 );

    return new echopoint.google.chart.LineChart(
    {
      renderId: "echopointUnitTestStrutComplexLineChart",
      styleName: "ComplexChart",
      data: data,
      title: title,
      axisLabels: labels,
      labelPositions: positions,
      lineStyles: lineStyles,
      rangeMarkers: ranges,
      fillArea: fillArea
    });
  }
});

File: src/server-java/test/echopoint/LineChartTest.java
/*
 * This file is part of the Echo Point Project.  This project is a
 * collection of Components that have extended the Echo Web Application
 * Framework Version 3.
 *
 * Version: MPL 1.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 */

package echopoint;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

import nextapp.echo.app.Component;
import nextapp.echo.app.Extent;
import nextapp.echo.app.Insets;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import echopoint.google.chart.LineChart;
import echopoint.google.chart.model.ChartData;
import echopoint.google.chart.model.FillArea;
import echopoint.google.chart.model.LineStyle;
import echopoint.google.chart.model.Range;
import echopoint.google.chart.model.RangeMarker;
import echopoint.google.chart.model.ShapeMarker;
import echopoint.google.chart.model.Title;

/**
 * Unit test suite for the {@link echopoint.google.chart.LineChart} component
 * wrapper.
 *
 * @author Rakesh Vidyadharan 2008-08-10
 * @version $Id: LineChartTest.java,v 1.1.1.1 2010-04-01 09:47:30 perxi Exp $
 */
public class LineChartTest extends GoogleChartTest<LineChart<Integer>>
{
  @BeforeClass
  public static void init()
  {
    set( new LineChart<Integer>() );
    setData( new ChartData<Integer>() );
  }

  @Test
  public void renderId()
  {
    final String id = "echopointUnitTestSimpleLineChart";
    getComponent().setRenderId( id );
    assertEquals( "Ensuring render id is same", id, getComponent().getRenderId() );
  }

  @Test
  public void chartData()
  {
    final Integer[] array = new Integer[] { 30,60,70,95,110,80 };
    final List<Integer> xdata = Arrays.asList( array );
    final int xmax = 120;

    getData().setXdata( xdata );
    getData().setXmax( xmax );
  }

  @Test
  public void markers()
  {
    final ArrayList<ShapeMarker> markers = new ArrayList<ShapeMarker>();
    markers.add( new ShapeMarker( "o", "ff3333", 5 ) );
    getData().setMarkers( markers );

    assertFalse( "Ensuring markers set", getData().getMarkers().isEmpty() );
  }

  @Test
  public void color()
  {
    final String color = "00ff00";
    getData().setColor( color );
    assertEquals( "Ensuring color set", color, getData().getColor() );
  }

  @Test
  public void title()
  {
    final Title title = new Title();
    title.add( "Simple Chart" );
    getComponent().setTitle( title );

    assertEquals( "Ensuring title set", title, getComponent().getTitle() );
  }

  @Test
  public void axisLabels()
  {
    final Collection<Collection<String>> labels =
        new ArrayList<Collection<String>>();

    String[] one =  new String[] { "0", "20", "40", "60", "80", "100" };
    labels.add( Arrays.asList( one ) );

    //String[] two = new String[] { "0", "25", "50", "75", "100", "125" };
    String[] two = new String[] { "Min", "Third", "Three Quarter", "Max" };
    labels.add( Arrays.asList( two ) );

    getComponent().setAxisLabels( labels );
    assertNotNull( "Ensuring labels set", getComponent().getAxisLabels() );
  }

  @Test
  public void labelPositions()
  {
    final Collection<Collection<Integer>> positions =
        new ArrayList<Collection<Integer>>();
    final Integer[] one = new Integer[] {};
    positions.add( Arrays.asList( one ) );

    final Integer[] two = new Integer[] { 1, 33, 75, 100 };
    positions.add( Arrays.asList( two ) );

    getComponent().setLabelPositions( positions );
    assertNotNull( "Ensuring positions set", getComponent().getAxisLabels() );
  }

  @Test
  public void axisRanges()
  {
    final Collection<Range> ranges = new ArrayList<Range>();
    ranges.add( new Range( 20, 125 ) );
    ranges.add( new Range( 25, 150 ) );

    getComponent().setAxisRanges( ranges );
    assertNotNull( "Ensuring axis ranges set", getComponent().getAxisRanges() );
  }

  @Test
  public void axisType()
  {
    final String type = "x,y";
    getComponent().setAxisType( type );
    assertEquals( "Ensuring axis type set", getComponent().getAxisType(), type );
  }

  @Test
  public void axisStyles()
  {
    final String style = "0,00ff33,13,1|1,0033ff,13,-1";
    getComponent().setAxisStyles( style );
    assertEquals( "Ensuring axis type set", getComponent().getAxisStyles(), style );
  }

  @Test
  public void rangeMarkers()
  {
    final Collection<RangeMarker> markers = new ArrayList<RangeMarker>();
    markers.add( new RangeMarker( "r", "ff0000", 0.1, 0.11 ) );
    markers.add( new RangeMarker( "R", "0000ff", 0.1, 0.11 ) );

    getComponent().setRangeMarkers( markers );
    assertNotNull( "Ensuring range markers set", getComponent().getRangeMarkers() );
  }

  @Test
  public void lineStyles()
  {
    final Collection<LineStyle> styles = new ArrayList<LineStyle>();
    styles.add( new LineStyle( 3, 6, 3 ) );

    getComponent().setLineStyles( styles );
    assertNotNull( "Ensuring line styles set", getComponent().getLineStyles() );
  }

  @Test
  public void fill()
  {
    final String fill = "bg,s,efefef|c,s,00000080";
    getComponent().setFill( fill );
    assertEquals( "Ensure fill set", fill, getComponent().getFill() );
  }

  @Test
  public void fillArea()
  {
    final Collection<FillArea> areas = new ArrayList<FillArea>();
    areas.add( new FillArea( "B", "224499", 0, 0 ) );

    getComponent().setFillArea( areas );
    assertNotNull( "Ensuring fill area set", getComponent().getFillArea() );
  }

  @Test
  public void gridLines()
  {
    final String grid = "10,10,1,0";
    getComponent().setGridLines( grid );
    assertEquals( "Ensure grid lines set", grid, getComponent().getGridLines() );
  }

  @Test
  public void insets()
  {
    final Insets insets = new Insets( new Extent( 10 ) );
    getComponent().setInsets( insets );
    assertEquals( "Ensure insets set", insets, getComponent().getInsets() );
  }

  @Test
  public void height()
  {
    final Extent height = new Extent( 400 );
    getComponent().setHeight( height );
    assertEquals( "Ensure height set", height, getComponent().getHeight() );
  }

  @Test
  public void width()
  {
    final Extent width = new Extent( 600 );
    getComponent().setWidth( width );
    assertEquals( "Ensure width set", width, getComponent().getWidth() );
  }

  @AfterClass
  @SuppressWarnings( {"unchecked"} )
  public static void finish()
  {
    final Component content = Application.getContent().getTestArea();
    content.removeAll();
    final LineChart<Integer> chart = (LineChart<Integer>) get();

    final ArrayList<ChartData<Integer>> collection = new ArrayList<ChartData<Integer>>();
    collection.add( getData() );
    chart.setData( collection );
    assertNotNull( "Ensuring that getData() is set", chart.getData() );

    content.add( get() );
  }
}
